File: frontend/src/qa/types/certification.ts
export interface Certification {
  id: string;
  certificate_number: string;
  name: string;
  type: 'ISO_9001' | 'ISO_14001' | 'ISO_45001' | 'TOURISM_QUALITY' | 'SAFETY_CERTIFICATION' | 'DRIVER_LICENSE' | 'GUIDE_LICENSE' | 'BUSINESS_LICENSE' | 'TRANSPORT_PERMIT' | 'OTHER';
  issuing_body: string;
  issuing_authority?: string;
  accreditation_body?: string;
  scope: 'COMPANY_WIDE' | 'DEPARTMENT' | 'INDIVIDUAL' | 'VEHICLE' | 'LOCATION';
  entity_type?: string;
  entity_id?: string;
  entity_name?: string;
  issue_date: string;
  expiry_date?: string;
  effective_date?: string;
  status: 'ACTIVE' | 'EXPIRED' | 'SUSPENDED' | 'PENDING_RENEWAL' | 'CANCELLED';
  document_path?: string;
  document_url?: string;
  verification_url?: string;
  requirements_met?: string;
  conditions?: string;
  restrictions?: string;
  renewable: boolean;
  renewal_process?: string;
  renewal_cost?: number;
  renewal_lead_time_days?: number;
  last_audit_date?: string;
  next_audit_date?: string;
  compliance_verified: boolean;
  certificate_holder?: string;
  responsible_manager?: string;
  description?: string;
  notes?: string;
  created_at: string;
  updated_at?: string;
  is_expired: boolean;
  days_until_expiry?: number;
  needs_renewal: boolean;
  is_valid: boolean;
  validity_period_days?: number;
  renewal_start_date?: string;
}

export interface CreateCertificationData {
  certificate_number: string;
  name: string;
  type: 'ISO_9001' | 'ISO_14001' | 'ISO_45001' | 'TOURISM_QUALITY' | 'SAFETY_CERTIFICATION' | 'DRIVER_LICENSE' | 'GUIDE_LICENSE' | 'BUSINESS_LICENSE' | 'TRANSPORT_PERMIT' | 'OTHER';
  issuing_body: string;
  issuing_authority?: string;
  accreditation_body?: string;
  scope: 'COMPANY_WIDE' | 'DEPARTMENT' | 'INDIVIDUAL' | 'VEHICLE' | 'LOCATION';
  entity_type?: string;
  entity_id?: string;
  entity_name?: string;
  issue_date: string;
  expiry_date?: string;
  effective_date?: string;
  requirements_met?: string;
  conditions?: string;
  restrictions?: string;
  renewable?: boolean;
  renewal_process?: string;
  renewal_cost?: number;
  renewal_lead_time_days?: number;
  certificate_holder?: string;
  responsible_manager?: string;
  description?: string;
  notes?: string;
}

export interface CertificationFilters {
  page?: number;
  size?: number;
  type?: string;
  status?: string;
  scope?: string;
  expiring_soon?: boolean;
  needs_renewal?: boolean;
}

export interface CertificationsResponse {
  items: Certification[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface RenewCertificationData {
  new_certificate_number: string;
  new_issue_date: string;
  new_expiry_date?: string;
  renewal_cost?: number;
  renewal_notes?: string;
}

export interface QADashboard {
  audit_summary: {
    total_audits: number;
    completed_this_month: number;
    pass_rate: number;
    overdue_audits: number;
  };
  nonconformity_summary: {
    total_open: number;
    critical_open: number;
    overdue_actions: number;
    resolved_this_month: number;
  };
  compliance_summary: {
    total_requirements: number;
    compliant: number;
    non_compliant: number;
    pending: number;
    expiring_soon: number;
  };
  certification_summary: {
    total_certifications: number;
    active: number;
    expiring_soon: number;
    expired: number;
  };
}

File: backend/qa_service/models/certification.py
"""
Certification model – QA service
"""
from __future__ import annotations

import uuid
from datetime import datetime, date, timedelta
from enum import Enum
from typing import Optional

from sqlmodel import Field, SQLModel


# ──────────────────────────────── ENUMS ────────────────────────────────
class CertificationType(str, Enum):
    ISO_9001 = "ISO 9001"
    ISO_14001 = "ISO 14001"
    ISO_45001 = "ISO 45001"
    TOURISM_QUALITY = "Tourism Quality"
    SAFETY_CERTIFICATION = "Safety Certification"
    DRIVER_LICENSE = "Driver License"
    GUIDE_LICENSE = "Guide License"
    BUSINESS_LICENSE = "Business License"
    TRANSPORT_PERMIT = "Transport Permit"
    OTHER = "Other"


class CertificationStatus(str, Enum):
    ACTIVE = "Active"
    EXPIRED = "Expired"
    SUSPENDED = "Suspended"
    PENDING_RENEWAL = "Pending Renewal"
    PENDING = "Pending"  # ↞ needed by service.create_certification


class CertificationScope(str, Enum):
    COMPANY_WIDE = "Company-wide"
    DEPARTMENT = "Department"
    INDIVIDUAL = "Individual"
    VEHICLE = "Vehicle"
    LOCATION = "Location"


class EntityType(str, Enum):
    """What object does the certificate bind to?"""
    COMPANY = "Company"
    DEPARTMENT = "Department"
    EMPLOYEE = "Employee"
    DRIVER = "Driver"
    VEHICLE = "Vehicle"
    LOCATION = "Location"
    OTHER = "Other"


# ──────────────────────────────── MODEL ────────────────────────────────
class Certification(SQLModel, table=True):
    __tablename__ = "certifications"

    # Keys
    id: Optional[uuid.UUID] = Field(default_factory=uuid.uuid4, primary_key=True)
    certificate_number: str = Field(unique=True, max_length=100, index=True)

    # Core data
    name: str = Field(max_length=255, index=True)
    type: CertificationType = Field(index=True)
    scope: CertificationScope = Field(index=True)

    # Binding
    entity_type: Optional[EntityType] = Field(default=None, index=True)
    entity_id: Optional[str] = Field(default=None, max_length=100, index=True)
    entity_name: Optional[str] = Field(default=None, max_length=255)

    # Issuing bodies
    issuing_body: str = Field(max_length=255)
    issuing_authority: Optional[str] = Field(default=None, max_length=255)
    accreditation_body: Optional[str] = Field(default=None, max_length=255)

    # Validity
    issue_date: date = Field(index=True)
    expiry_date: Optional[date] = Field(default=None, index=True)
    effective_date: Optional[date] = Field(default=None)
    status: CertificationStatus = Field(default=CertificationStatus.ACTIVE, index=True)

    # Docs
    document_path: Optional[str] = Field(default=None, max_length=500)
    document_url: Optional[str] = Field(default=None, max_length=500)
    verification_url: Optional[str] = Field(default=None, max_length=500)

    # Renewal / audit
    renewable: bool = Field(default=True)
    renewal_process: Optional[str] = Field(default=None, max_length=1000)
    renewal_cost: Optional[float] = Field(default=None, ge=0)
    renewal_lead_time_days: Optional[int] = Field(default=None, ge=0)
    renewal_notes: Optional[str] = Field(default=None, max_length=1000)
    last_audit_date: Optional[date] = Field(default=None)
    next_audit_date: Optional[date] = Field(default=None)

    # Suspension
    suspension_reason: Optional[str] = Field(default=None, max_length=1000)

    # Book-keeping
    compliance_verified: bool = Field(default=True)
    created_by: Optional[uuid.UUID] = Field(default=None, index=True)
    responsible_manager: Optional[uuid.UUID] = Field(default=None)
    description: Optional[str] = Field(default=None, max_length=1000)
    notes: Optional[str] = Field(default=None, max_length=2000)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: Optional[datetime] = Field(default=None)

    # ─────────────── Helper methods ───────────────
    def is_expired(self) -> bool:
        return bool(self.expiry_date and date.today() > self.expiry_date)

    def days_until_expiry(self) -> Optional[int]:
        return None if not self.expiry_date else (self.expiry_date - date.today()).days

    def needs_renewal(self, alert_days: int = 60) -> bool:
        return self.renewable and self.days_until_expiry() is not None and \
            self.days_until_expiry() <= alert_days

    def is_valid(self) -> bool:
        return self.status == CertificationStatus.ACTIVE and \
            not self.is_expired() and self.compliance_verified

    def get_validity_period_days(self) -> Optional[int]:
        if not self.expiry_date:
            return None
        start = self.effective_date or self.issue_date
        return (self.expiry_date - start).days

    def calculate_renewal_start_date(self) -> Optional[date]:
        if not (self.renewable and self.expiry_date and self.renewal_lead_time_days):
            return None
        return self.expiry_date - timedelta(days=self.renewal_lead_time_days)
